core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2219,11 +2219,46 @@ object ZIO {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    val n         = as.size
+    val resultArr = new Array[Any](n)
+    val resultList: ZIO[R, E, List[B]] = for {
+      parentId <- ZIO.fiberId
+      latch    <- CountdownLatch.make(n)
+      failed   <- Promise.make[Nothing, Unit]
+      cause    <- Ref.make[Cause[E]](Cause.empty)
+      fibers <- as
+                 .foldLeft[(Int, URIO[R, List[Fiber[E, _]]])]((0, URIO.succeed(Nil))) {
+                   case ((i, acc), a) =>
+                     val task = ZIOFn(fn)(
+                       (x: A) =>
+                         fn(x).traced
+                           .foldCauseM(
+                             c => cause.update(cs => cs && c) *> failed.succeed(()).unit,
+                             b => ZIO.effectTotal(resultArr(i) = b)
+                           )
+                           .ensuring(latch.countDown)
+                           .untraced
+                     )
+
+                     val appended = for {
+                       fs <- acc
+                       f  <- task(a).fork
+                     } yield f :: fs
+
+                     (i + 1, appended)
+                 }
+                 ._2
+      interrupter = (failed.await *>
+        ZIO.foreach(fibers)(_.interruptAs(parentId))) // The failed fiber can be safely interrupted.
+      .toManaged_.fork
+      _             <- interrupter.use_(latch.await)
+      combinedCause <- cause.get
+      results <- if (combinedCause.isEmpty) UIO.succeed(resultArr.asInstanceOf[Array[B]].toList)
```

Comment:
If I'm not mistaken, updates to `resultArr` won't always be visible here, because there is no synchronization on its value. So, there is a race condition.

Reply:
The access to resultArr there is guarded by the latch. The completion of the effect on line 2255 guarantees that all of the fibers that were spawned have finished (in one way or another) and so no further changes can be made to the array. Unless there is another way this point could be reached that I'm missing?

Reply:
From the point of view of Java Memory Model, is it guaranteed, that finishing of these fibers happens-before accessing `resultArr` on L2257? I'm not quite sure about this...

Reply:
Thinking about it a bit more, you're absolutely correct. I have a couple of ideas of how to fix it but haven't had a chance to look at it this weekend.

Reply:
`write to array` HB `latch.countDown`
`latch.countDown` HB `latch.await`
`latch.await` HB `read from array`
is it wrong?

Reply:
Yeah, seems correct.

Reply:
I changed it to an AtomicReferenceArray which will make the reads and writes to the elements volatile which I _think_ should be adequate.